### PR TITLE
MKLDNN Subgraph Optimization

### PR DESCRIPTION
**Pull request by sreekanth-yalachigere**
This optimization creates a subgraph of mkldnn operators and uses full potential of mkldnn by using block propagation and improves performance significantly. 
For example, in Resnet50, currently we have 109 data reorders. 

```
onnxruntime_data(nchw) ->reorder(nchw8c)->conv(nchw8c) -> reorder(nchw);
onnxruntime_data(nchw) ->batchnorm(nchw);
onnxruntime_data(nchw) ->Relu(nchw);
onnxruntime_data(nchw) ->reorder(nchw8c)->pool(nchw8c) -> reorder(nchw);
onnxruntime_data(nchw) ->reorder(nchw8c)->conv(nchw8c) -> reorder(nchw);
```

With this optimization, we will have only one data reorder and we propage blocked mkldnn memory all the way to the end of subgraph.

`onnxruntime_data(nchw) -> conv(nchw8c) -> batchnorm-relu(nchw8c)->pool(nchw8c)->conv(nchw8c)... reorder(nchw)`

Subgraph optimization can be enabled by setting the following environment variable.
`ORT_MKLDNN_SUBGRAPH=1`

### CODE REVIEW

**Comment by snnn** [comment]
Which Resnet50 model are you testing?

**Comment by sreekanth-yalachigere** [comment]
@snnn master: https://s3.amazonaws.com/download.onnx/models/opset_9/resnet50.tar.gz from model-zoo.

**Comment by snnn** [comment]
> @snnn master: https://s3.amazonaws.com/download.onnx/models/opset_9/resnet50.tar.gz from model-zoo.

That model comes from caffe2.  On that model, onnxruntime performs very bad when MKLDNN or MKLML is enabled.

Below are metrics from my local dev machine.
\ |with MKLML | without MKLML
--|-- | --
model from TF | 31ms | 31ms
model from caffe2 | 223ms | 36ms

File: onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  std::shared_ptr<Subgraph> subgraph;
+  std::shared_ptr<MKLContext> mkl_context;
+
+  SubgraphParams() {}
+};
+
+template <typename T>
+class SubgraphPrimitive : public PrimitiveBase {
+ public:
+  explicit SubgraphPrimitive(const ONNXRunTimeTensor* input_tensors,
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h]: You don't need the explicit keyword here.

File: onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+      Initialize(input_tensors);
+    }
+  }
+
+  void CreateKernels(const SubgraphParams& params) {
+    for (auto& mklnode : params.subgraph->mklnodes) {
+      if (mklnode.name == "Conv") {
+        std::ostringstream os;
+        os << "Conv-" << mklnode.node_index << "-";
+        std::shared_ptr<MklConv<T>> kernel;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h]: Do all these have to be shared_ptrs?

File: onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc [hunk 3]
```diff
@@ -107,6 +112,263 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
...
 
+std::vector<std::unique_ptr<ComputeCapability>> MKLDNNExecutionProvider::GetCapability(
+    const onnxruntime::GraphViewer& graph_viewer,
+    const std::vector<const KernelRegistry*>& kernel_registries) const {
+  ORT_UNUSED_PARAMETER(kernel_registries);
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+
+  // switch between mkldnn-vanilla and mkldnn-subgraph implementation using
+  // MKLDNN_SUBGRAPH environment variable
+  const char* env = getenv("ORT_MKLDNN_SUBGRAPH");
```
[the remark this answers is not in the document]
Comment (jywu-msft) [reply onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: this is a temp switch. didn't want to pollute session/run options with this.
once the behavior is default, there is no longer a need for it.

[the remark this answers is not in the document]
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: I will update the comment.

File: onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc [hunk 4]
```diff
@@ -107,6 +112,263 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
 }
 }  // namespace mkl_dnn
 
+std::vector<std::unique_ptr<ComputeCapability>> MKLDNNExecutionProvider::GetCapability(
```
[the remark this answers is not in the document]
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: Yes. I did create CreateMetaDef but it's still long.  I will start with extracting switch between vanilla and subgraph.

File: onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h (hunk 2 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h]: Yes. Each kernel has pointers to parent kernels to get previous node's output and memory format.

File: onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc [hunk 5]
```diff
@@ -107,6 +112,263 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
...
+        for (auto attIt = attributes.begin(); attIt != attributes.end(); ++attIt) {
+          std::string key = node->OpType() + "-" + std::to_string(index) + "-" + attIt->first;
+          std::pair<std::string, ONNX_NAMESPACE::AttributeProto> att(key, attIt->second);
+          subgraph_attributes[key] = attIt->second;
+        }
+      }
+
+      auto temp_index = node_index + 1;
+      if (temp_index < graph_viewer.MaxNodeIndex()) {
+        if (sub_var.subgraph_node_indexes.size() > 0) {
```
[the remark this answers is not in the document]
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: corrected.

File: onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: Created new private function RunSubgraph.

File: onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc [hunk 6]
```diff
@@ -107,6 +111,276 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
...
+void MKLDNNExecutionProvider::CreateMetaDef(SubgraphVariables& sub_var, const NodeAttributes& subgraph_attributes,
+                                            std::vector<std::unique_ptr<ComputeCapability>>& result) const {
+  std::string graph_fused_nodes;
+  std::string node_list;
+  std::string subgraph_id = std::to_string(sub_var.subgraph_index);
+  sub_var.subgraph_index++;
+
+  auto meta_def = std::make_unique<::onnxruntime::IndexedSubGraph::MetaDef>();
+  meta_def->name = "MkldnnCustomOp" + std::to_string(sub_var.subgraph_index);
+  meta_def->domain = kMklDnnExecutionProvider;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: this should be domain of the op implementation. put kMSDomain for now (meaning custom implementation)

File: onnxruntime/core/providers/mkldnn/subgraph/MklBatchNorm.h [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+                                       const TensorShape& scale_shape,
+                                       const TensorShape& b_shape,
+                                       const TensorShape& mean_shape,
+                                       const TensorShape& var_shape) {
+    // defined as per spec and used for validation
+    constexpr int kNumInputScaleDimensions = 1;
+    constexpr int kNumInputBiasDimensions = 1;
+    constexpr int kNumInputMeanDimensions = 1;
+    constexpr int kNumInputVarianceDimensions = 1;
+    //constexpr int kMinCudaNumDims = 4;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/subgraph/MklBatchNorm.h]: remove

File: onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                                  mkldnn::engine& cpu_engine,
+                                  std::vector<mkldnn::primitive>& net,
+                                  mkldnn::memory::format& src_fmt) = 0;
+
+  virtual void ReorderWeights(const ONNXRunTimeTensor* input_tensors, mkldnn::engine& cpu_engine) {
+    ORT_UNUSED_PARAMETER(input_tensors);
+    ORT_UNUSED_PARAMETER(cpu_engine);
+  }
+
+  virtual Status Compute(const ONNXRunTimeTensor* input_tensors,
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h]: i don't think we should call this Compute() 
it's not doing any computation.

File: onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+
+  static SubgraphPrimitivePool& GetInstance() {
+    static SubgraphPrimitivePool pool;
+    return pool;
+  }
+};
+}  // namespace
+
+template <typename T>
+class MkldnnCustomOp {
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h]: i think we can find another name that is more intuitive and describes the functionality.
will discuss with @pranavsharma

File: onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h (hunk 8 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h]: Does Submit work?

File: onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc (hunk 6 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: changed to kMSDomain.

File: onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h (hunk 8 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h]: Renamed it to Submit. Here we are submitting data to MKLDNN primitives.

Comment (jywu-msft) [reply onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h]: I think in a previous discussion we talked about "Bind" ?  binding inputs and outputs?

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h]: Yes. I missed it. Renamed it to Bind.

**Comment by pranavsharma** [comment]
> I'm not fully done with the review. Will get back soon.

starting to look again. i see that a lot more files got added as a result of merging with master. difficult to see the diff. most of the changes are in the mkldnn folder, so should be easy to call out mostly.

**Comment by sreekanth-yalachigere** [comment]
Many files hare changed because of master merge, Closing this and creating a new PR.

**Pull request closed without merging**